src/Microsoft.AspNetCore.Server.Testing/Deployers/NanoDeployer/NanoDeployer.cs

```diff
@@ -0,0 +1,251 @@
+            {
+                throw new InvalidOperationException($"Expected the deployment parameters of type {nameof(NanoDeploymentParameters)}.");
+            }
+
+            if (_deploymentParameters.ServerType != ServerType.IIS
+                && _deploymentParameters.ServerType != ServerType.Kestrel
+                && _deploymentParameters.ServerType != ServerType.WebListener)
+            {
+                throw new InvalidOperationException($"Server type {_deploymentParameters.ServerType} is not supported for Nano." +
+                    " Supported server types are Kestrel, IIS and WebListener");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.ServerName))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.ServerName)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.ServerAccountName))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.ServerAccountName)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.ServerAccountPassword))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.ServerAccountPassword)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.NanoServerFileShare))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.NanoServerFileShare)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.NanoServerRelativeExecutablePath))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.NanoServerRelativeExecutablePath)}");
+            }
+        }
+
+        public override DeploymentResult Deploy()
+        {
+            // Publish the app to a local temp folder on the machine where the test is running
+            DotnetPublish();
+
+            var folderId = Guid.NewGuid().ToString();
+            _deployedFolderPathInFileShare = Path.Combine(_deploymentParameters.NanoServerFileShare, folderId);
+
+            DirectoryCopy(
+                _deploymentParameters.PublishedApplicationRootPath,
+                _deployedFolderPathInFileShare,
+                copySubDirs: true);
+            Logger.LogInformation($"Copied the locally published folder to the file share path '{_deployedFolderPathInFileShare}'");
+
+            // Copy the scripts from this assembly's embedded resources to the temp path on the machine where these
+            // tests are being run
+            var embeddedFileProvider = new EmbeddedFileProvider(
+                GetType().GetTypeInfo().Assembly,
+                "Microsoft.AspNetCore.Server.Testing.Deployers.NanoDeployer");
+
+            var filesOnDisk = CopyEmbeddedScriptFilesToDisk(
```

Comment:
Static/singleton? There's no point to read these everytime this method is executed